### PR TITLE
Support latest version while upgrading VM hardware.

### PR DESCRIPTION
**Pull request by pgbidkar**
Fix issue https://github.com/ansible/ansible/issues/56273

##### SUMMARY
Change support latest version while upgrading VM hardware

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
vmware_guest.py

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 1]
```diff
@@ -151,6 +151,8 @@
     - ' - C(cpu_reservation) (integer): The amount of CPU resource that is guaranteed available to the virtual machine.
           Unit is MHz. version_added: 2.5'
     - ' - C(version) (integer): The Virtual machine hardware versions. Default is 10 (ESXi 5.5 and onwards).
+          If value specified as C(latest), version is set to the most current virtual hardware supported on the host.
+          C(latest) is added in version 2.9.
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]:
```suggestion
          C(latest) is added in version 2.10.
```

File: test/integration/targets/vmware_guest/tasks/reconfig_vm_to_latest_version.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Test code for the vmware_guest module.
+# Copyright: (c) 2019, Pavan Bidkar <pbidkar@vmware.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+- name: Upgrade VM to latest version
```
Comment (Akasurde) [inline test/integration/targets/vmware_guest/tasks/reconfig_vm_to_latest_version.yml:6]: Can you please again run this same task for check idempotency ?

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 1 above)
Comment (pgbidkar) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: done

File: test/integration/targets/vmware_guest/tasks/reconfig_vm_to_latest_version.yml (hunk 2 above)
Comment (pgbidkar) [reply test/integration/targets/vmware_guest/tasks/reconfig_vm_to_latest_version.yml:6]: Done

File: test/integration/targets/vmware_guest/tasks/reconfig_vm_to_latest_version.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  vmware_guest:
+    validate_certs: False
+    hostname: "{{ vcenter_hostname }}"
+    username: "{{ vcenter_username }}"
+    password: "{{ vcenter_password }}"
+    name: "{{ item.name }}"
```
Comment (goneri) [inline test/integration/targets/vmware_guest/tasks/reconfig_vm_to_latest_version.yml]: `item` is only defined when you are  in a loop context. You can use `test_vm1` here instead.

Comment (pgbidkar) [reply test/integration/targets/vmware_guest/tasks/reconfig_vm_to_latest_version.yml]: reverted

**Comment by pgbidkar** [comment]
Hi @goneri @jillr , Please can you check and help in fixing precheckin failures?https://app.shippable.com/github/ansible/ansible/runs/143329/137/console. check for idempotency  is failing.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 4]
```diff
@@ -39,7 +39,7 @@
     - "   Resource > Assign Virtual Machine to Resource Pool"
     - "Module may require additional privileges as well, which may be required for gathering facts - e.g. ESXi configurations."
     - Tested on vSphere 5.5, 6.0, 6.5 and 6.7
-    - Use SCSI disks instead of IDE when you want to expand online disks by specifying a SCSI controller
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: This was spelled correctly originally.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 5]
```diff
@@ -165,12 +167,12 @@
     - This parameter is case sensitive.
     - 'Examples:'
     - "  virtual machine with RHEL7 64 bit, will be 'rhel7_64Guest'"
-    - "  virtual machine with CentOS 64 bit, will be 'centos64Guest'"
+    - "  virtual machine with CensOS 64 bit, will be 'centos64Guest'"
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: As was this.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 6]
```diff
@@ -295,7 +297,7 @@
     description:
     - A list of networks (in the order of the NICs).
     - Removing NICs is not allowed, while reconfiguring the virtual machine.
-    - All parameters and VMware object names are case sensitive.
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: Here too.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 7]
```diff
@@ -782,7 +784,7 @@ def integer_value(self, input_value, name):
         """
         Function to return int value for given input, else return error
         Args:
-            input_value: Input value to retrieve int value from
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: And here.

**Review by jillr — COMMENTED** [review]
It looks like the new tests might have exposed an existing bug in `vmware_guest.py`.  If I run your test suite (with `version: 14`) on `devel` the idempotency test still fails.  Something else in the module is setting `self.change_applied = True` before you use it at line 1299.  
I would start by looking at whether or not this section of code is being reached, as it's the most obvious place to me this attribute could be set.
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py#L2543-L2563

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 4 above)
Comment (pgbidkar) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Oops! sorry. This is not my change. I think it got escaped in my PR after I rebased.
I will revert this!

**Comment by pgbidkar** [comment]
> It looks like the new tests might have exposed an existing bug in `vmware_guest.py`. If I run your test suite (with `version: 14`) on `devel` the idempotency test still fails. Something else in the module is setting `self.change_applied = True` before you use it at line 1299.
> I would start by looking at whether or not this section of code is being reached, as it's the most obvious place to me this attribute could be set.
> https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py#L2543-L2563

Thanks @jillr , Yes I did came across this section where self.change_applied is set. But If I look at both the conditions, 
```if self.params['is_template'] and not self.current_vm_obj.config.template:``` & ```elif not self.params['is_template'] and self.current_vm_obj.config.template:```

Both will be False in my playbook. I will re check the flow

Thanks

**Comment by pgbidkar** [comment]
> It looks like the new tests might have exposed an existing bug in `vmware_guest.py`. If I run your test suite (with `version: 14`) on `devel` the idempotency test still fails. Something else in the module is setting `self.change_applied = True` before you use it at line 1299.
> I would start by looking at whether or not this section of code is being reached, as it's the most obvious place to me this attribute could be set.
> https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py#L2543-L2563

Hi @jillr , skipping idepotency test for now for this fix. 
There is bug in reconfigure_vm() method. self.changeapplied variable not correctly used. I will raise new issue and work on it. 
Thanks

**Review by goneri — APPROVED** [review]
Looks good to me, but I would prefer you open the issue before we land the patch.

**Comment by pgbidkar** [comment]
> Looks good to me, but I would prefer you open the issue before we land the patch.

Thanks @goneri , I have opened the issue https://github.com/ansible/ansible/issues/62408

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 8]
```diff
@@ -1246,42 +1248,55 @@ def configure_hardware_params(self, vm_obj):
             if 'version' in self.params['hardware']:
                 hw_version_check_failed = False
                 temp_version = self.params['hardware'].get('version', 10)
-                try:
-                    temp_version = int(temp_version)
-                except ValueError:
-                    hw_version_check_failed = True
+                if temp_version.lower() == 'latest':
```
Comment (goneri) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1251]: There is already way too much if in this part of the code, may I suggest you split up the logic in a new function?

File: changelogs/fragments/62188-VMware-Guest-Support-latest-version-while-upgrading-VM-hardware.yml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+- Support latest version while upgrading VM hardware. Issue #56273
```
Comment (goneri) [inline changelogs/fragments/62188-VMware-Guest-Support-latest-version-while-upgrading-VM-hardware.yml]: Please, clarify it's a vmware feature.

**Review by goneri — CHANGES_REQUESTED** [review]
Thanks @pgbidkar, could you just fix this little problem, or it will be confusing for the users.

File: changelogs/fragments/62188-VMware-Guest-Support-latest-version-while-upgrading-VM-hardware.yml (hunk 9 above)
Comment (pgbidkar) [reply changelogs/fragments/62188-VMware-Guest-Support-latest-version-while-upgrading-VM-hardware.yml]: Done

File: changelogs/fragments/62188-VMware-Guest-Support-latest-version-while-upgrading-VM-hardware.yml [hunk 10]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+- VMWare Support latest version while upgrading Virtual Machine hardware. It is VMWare feature and latest param will upgrade to latest hardware version Issue #56273.
```
Comment (goneri) [inline changelogs/fragments/62188-VMware-Guest-Support-latest-version-while-upgrading-VM-hardware.yml]: I still don't like it :-), I suggest:

- vmware_guest - Add ability to upgrade the guest hardware version to ``latest``.

Comment (pgbidkar) [reply changelogs/fragments/62188-VMware-Guest-Support-latest-version-while-upgrading-VM-hardware.yml]: Thanks @goneri 
Done the changes. Please review

**Review by goneri — APPROVED** [review]

**Comment by DomFilteau** [comment]
Hi, 
ESXi 6.7 Update 2 is actually Hardware version 15, 
Line 1254 check should be changed to include version 15, which is the current latest..
Regards

**Comment by goneri** [comment]
@pgbidkar can you do this little change before we merge the module?

**Comment by pgbidkar** [comment]
> Hi,
> ESXi 6.7 Update 2 is actually Hardware version 15,
> I think that line 1254 check should be changed to include version 15, which is the current latest..
> Regards

Thanks for pointing this. fixed it

**Pull request merged**